compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/RangeContainsLowering.kt

```diff
@@ -0,0 +1,444 @@
+        parameterCount { it == 1 }
+    }
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        // The call to contains() in `5 in 0..10` looks like this in IR:
+        //
+        //   CALL 'public open fun contains (value: kotlin.Int): kotlin.Boolean [operator] declared in kotlin.ranges.IntRange' type=kotlin.Boolean origin=IN
+        //
+        // And when `!in` is used in `5 !in 0..10`:
+        //
+        //   CALL 'public final fun not (): kotlin.Boolean [operator] declared in kotlin.Boolean' type=kotlin.Boolean origin=NOT_IN
+        //     $this: CALL 'public open fun contains (value: kotlin.Int): kotlin.Boolean [operator] declared in kotlin.ranges.IntRange' type=kotlin.Boolean origin=NOT_IN
+        //
+        // We only want to lower the call to contains(); even in the `!in` case, the call to not() should be preserved).
+        val origin = expression.origin
+        if (origin == null || (origin != IrStatementOrigin.IN && origin != IrStatementOrigin.NOT_IN)) {
+            return super.visitCall(expression)  // The call is not an `in` expression.
+        }
+        if (origin == IrStatementOrigin.NOT_IN && expression.symbol == context.irBuiltIns.booleanNotSymbol) {
+            return super.visitCall(expression)
+        }
+
+        if (expression.extensionReceiver != null && !stdlibExtensionContainsCallMatcher(expression)) {
+            // We can only optimize calls to the stdlib extension functions and not a user-defined extension.
+            // TODO: This breaks the optimization for *Range.reversed().contains(). The function called there is the extension function
+            // Iterable.contains(). Figure out if we can safely match on that as well.
+            return super.visitCall(expression)
+        }
+
+        // The HeaderInfoBuilder extracts information (e.g., lower/upper bounds, direction) from the range expression, which is the
+        // receiver for the contains() call.
+        val receiver = expression.dispatchReceiver ?: expression.extensionReceiver!!
+        val headerInfo = receiver.accept(headerInfoBuilder, expression)
+            ?: return super.visitCall(expression)  // The receiver is not a supported range (or not a range at all).
+
+        val argument = expression.getValueArgument(0)!!
+
+        val builder = context.createIrBuilder(getScopeOwnerSymbol(), expression.startOffset, expression.endOffset)
+
+        return builder.run {
+            var replacement = buildContainsComparison(headerInfo, argument, origin)
+                ?: return super.visitCall(expression)  // The call cannot be lowered.
+            if (argument.type.isNullable()) {
+                // There are stdlib extension functions that return false for null arguments, e.g., IntRange.contains(Int?).
+                replacement = irIfNull(context.irBuiltIns.booleanType, argument.deepCopyWithSymbols(), irFalse(), replacement)
```

Comment:
Is it possible to avoid deep copy? For some reason it fails on a complex expression here:
```
fun foo(x: Int?, a: Int, b: Int): Boolean =
    { x }() in a..b

fun main() {
    foo(null, 0, 5)
}
```
```
java.lang.NullPointerException: declarationParentsStack.peekFirst() must not be null
	at org.jetbrains.kotlin.ir.util.PatchDeclarationParentsVisitor.patchParent(PatchDeclarationParents.kt:57)
	...
	at org.jetbrains.kotlin.backend.common.lower.Transformer.visitCall(RangeContainsLowering.kt:452)
```

BTW I was actually going to construct a test case related to https://youtrack.jetbrains.com/issue/KT-40467, which is the reason we currently mostly avoid `deepCopyWithSymbols` in JVM IR...

Reply:
Oh, looks like it also breaks semantics here:
```
var state = ""
fun sideEffect(): Int? {
    state += "OK"
    return 1
}

fun box(): String {
    sideEffect() in 1..2
    return state
}
```
returns "OKOK"

Reply:
I believe I fixed this now in latest version of PR (last commit) and added tests. Please take a look!

Reply:
Could you clarify why are we optimizing the case of nullable subject? Is it a pattern worth optimizing and were there any requests to optimize it?

I'm worried that the code of the lowering became much more involved, and the motivation is not clear to me. Maybe we could avoid optimizing these kinds of expressions, and force the user to refactor the code in a way that would make the subject non-nullable (`x != null && x in range`).

Reply:
I'm very glad you asked this question! There weren't any requests to do optimize this, and it is also not optimized on the old JVM backend. During development, one of the initial failures I saw was `nullableInPrimitiveRange.kt`, because I wasn't handling nullable subjects. I considered at first to just let it remain unoptimized and let the stdlib extension functions take care of it (e.g., `IntRange.contains(Int?)`), but I thought it was simple enough to handle, hence the initial version which was indeed very simple.

But after the problems you've helped identify, and the corresponding solution, it has become much more complicated to handle. It's great you asked the question since I was also getting a little wary of the complexity. I'll just punt on handling it for now.